### PR TITLE
[Test]Add Model URL Connectivity Tests and Improve URL Handling

### PR DESCRIPTION
**Pull request by weiyuanyue**
This PR introduces comprehensive testing infrastructure to validate model download URLs and refactors URL handling code for better maintainability, robustness, and testability.

### New Integration Test Suite: `ModelUrlConnectivityTests.cs`
The centerpiece of this PR is a robust test suite that automatically validates all model URLs defined in `.model.json` and `.modelgroup.json` files across the codebase. This test:

- **Prevents broken links**: Automatically detects when model download URLs become inaccessible
- **Validates real download flows**: Tests the actual URLs used during downloads (raw URLs for GitHub files, resolve URLs for HuggingFace files, API endpoints for directories)
- **Provides actionable feedback**: Clear error messages showing which models have broken URLs, status codes, and download endpoints
- **Supports both individual and batch testing**: Run tests per file or get a comprehensive report across all model definitions

**Business Value**: This test suite acts as an early warning system, helping developers quickly identify and fix broken model links before they impact users.

### 1. New Test Files
- **`ModelUrlConnectivityTests.cs`** (352 lines): Integration tests that make real HTTP requests to validate model URL accessibility
- **`ModelUrlTests.cs`** (589 lines): Comprehensive unit tests for URL parsing, construction, and edge cases without network calls
- **`ApiIntegrationTests.cs`** (101 lines): Integration tests for GitHub and HuggingFace API functionality
- **`ApiTests.cs`** (126 lines): Unit tests for API helper classes

### 2. Enhanced `ModelUrl.cs` Utility Class
**Before**: URLs were constructed inline with string concatenation throughout the codebase  
**After**: Centralized URL building methods with proper validation

New static builder methods added:
- `HuggingFaceUrl`:
  - `BuildRepoUrl(org, repo)` - Base repository URLs
  - `BuildTreeUrl(org, repo, ref, path)` - Directory browsing URLs
  - `BuildResolveUrl(org, repo, ref, filePath)` - File download URLs
  - `BuildBlobUrl(org, repo, ref, filePath)` - File viewing URLs
  - `BuildApiUrl(org, repo, ref, path)` - API endpoints

- `GitHubUrl`:
  - `BuildRepoUrl(org, repo)` - Base repository URLs
  - `BuildApiUrl(org, repo, ref, path)` - API endpoints
  - `BuildRawUrl(org, repo, ref, filePath)` - Raw file download URLs
  - `BuildBlobUrl(org, repo, ref, filePath)` - File viewing URLs

**Benefits**:
- ✅ Single source of truth for URL construction
- ✅ Consistent URL formatting across the codebase
- ✅ Input validation with descriptive error messages
- ✅ Easier to test and maintain
- ✅ Supports both display URLs and actual download URLs

### 3. Refactored `ModelInformationHelper.cs`
- Replaced inline URL string concatenation with new builder methods
- Improved code readability and maintainability
- Better null handling and validation
- Cleaner separation between display URLs and API URLs

### 4. Updated `AddHFModelView.xaml.cs`
- Replaced hardcoded URL construction with builder methods
- Added proper validation for HuggingFace model IDs (must be in `org/repo` format)
- Enhanced null safety and error handling
- More robust license URL handling

### 5. Improved API Helpers (`GithubApi.cs`, `HuggingFaceApi.cs`)
- Use centralized URL builders instead of local constants
- Added proper error handling with HTTP status code checks
- Better exception messages for debugging
- Input validation for file paths

### 6. Enhanced `ModelDownload.cs`
- Minor improvements aligned with URL handling changes

##  Testing Strategy

The test suite validates URLs at two levels:

1. **Unit Tests** (`ModelUrlTests.cs`, `ApiTests.cs`):
   - Fast, no network calls
   - Test URL parsing, construction, edge cases
   - Validate error handling for malformed URLs
   - Test local path mapping logic

2. **Integration Tests** (`ModelUrlConnectivityTests.cs`, `ApiIntegrationTests.cs`):
   - Real HTTP requests to validate accessibility
   - Test both individual files and batch processing
   - Verify actual download flow mirrors business logic
   - Rate-limited to be respectful to servers

##  Test Coverage

The connectivity tests cover:
-  All `.model.json` files
-  All `.modelgroup.json` files
-  GitHub file URLs (tests raw.githubusercontent.com)
-  GitHub directory URLs (tests API endpoints)
-  HuggingFace file URLs (tests resolve URLs)
-  HuggingFace directory URLs (tests API endpoints)

##  Future Benefits

1. **Proactive Monitoring**: Run these tests in CI/CD to catch broken links before deployment
2. **Easier Refactoring**: Centralized URL logic makes future changes safer
3. **Better Debugging**: Clear error messages help identify issues quickly
4. **Quality Assurance**: Ensures model definitions remain valid over time

##  Note

Integration tests make real HTTP requests and should be run separately from unit tests in CI/CD pipelines to avoid:
- Long test execution times
- Rate limiting issues
- Network dependency in core test suites

### CODE REVIEW

File: AIDevGallery/Utils/HuggingFaceApi.cs [hunk 1]
```diff
@@ -26,7 +23,7 @@ internal class HuggingFaceApi
     /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
     public static async Task<List<HFSearchResult>?> FindModels(string query, string filter = "onnx")
     {
-        string searchUrl = $"{HFApiUrl}/models?search={query}&filter={filter}&full=true&config=true";
+        string searchUrl = $"{HuggingFaceUrl.ApiUrl}/models?search={query}&filter={filter}&full=true&config=true";
         using var client = new HttpClient();
         var response = await client.GetAsync(searchUrl);
```
[the remark this answers is not in the document]
Comment (weiyuanyue) [reply AIDevGallery/Utils/HuggingFaceApi.cs:28]: resolved in IDispose related PR

File: AIDevGallery/Utils/ModelDownload.cs [hunk 2]
```diff
@@ -197,6 +197,13 @@ public override void CancelDownload()
 
         filesToDownload = ModelInformationHelper.FilterFiles(filesToDownload, Details.FileFilters);
 
+        if (filesToDownload.Count == 0)
+        {
+            ModelDownloadFailedEvent.Log(Details.Url, new InvalidOperationException("No files to download after filtering"));
+            DownloadStatus = DownloadStatus.Canceled;
```
[the remark this answers is not in the document]
Comment (weiyuanyue) [reply AIDevGallery/Utils/ModelDownload.cs:203]: We dont have DownloadStatus.Failed

File: AIDevGallery.Utils/ModelUrl.cs [hunk 3]
```diff
@@ -75,10 +89,17 @@ public string GetLocalPath(string cacheRoot)
 /// </summary>
 public class HuggingFaceUrl : ModelUrl
 {
+    internal const string BaseUrl = "https://huggingface.co";
+
+    /// <summary>
+    /// The base API URL for Hugging Face
+    /// </summary>
+    public const string ApiUrl = "https://huggingface.co/api";
```
Comment (haoliuu) [inline AIDevGallery.Utils/ModelUrl.cs:96]: should this also be ``internal``?

File: AIDevGallery.Utils/ModelUrl.cs [hunk 4]
```diff
@@ -146,12 +168,108 @@ public HuggingFaceUrl(string modelNameOrUrl)
...
+    /// <param name="organization">The organization name</param>
+    /// <param name="repo">The repository name</param>
+    /// <param name="ref">The branch or commit reference</param>
+    /// <param name="path">The path within the repository</param>
+    /// <returns>The tree URL</returns>
+    public static string BuildTreeUrl(string organization, string repo, string @ref = "main", string? path = null)
+    {
+        ThrowIfNullOrWhiteSpace(organization);
+        ThrowIfNullOrWhiteSpace(repo);
+        ThrowIfNullOrWhiteSpace(@ref);
```
Comment (haoliuu) [inline AIDevGallery.Utils/ModelUrl.cs]: nit: would ``ArgumentNullException.ThrowIfNullOrWhiteSpace()`` works better than the custom helper method here? I see the func in this PR has no special logic.

File: AIDevGallery.Tests/IntegrationTests/ModelUrlConnectivityTests.cs [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+
+        foreach (var file in modelFiles)
+        {
+            yield return new object[] { Path.GetFileName(file) };
+        }
+    }
+
+    [TestMethod]
+    [DynamicData(nameof(GetModelGroupFiles), DynamicDataSourceType.Method)]
+    public async Task ModelUrlsShouldBeAccessible(string fileName)
```
Comment (haoliuu) [inline AIDevGallery.Tests/IntegrationTests/ModelUrlConnectivityTests.cs:71]:
```suggestion
    [TestMethod]
    [TestCategory("Integration")]
    [TestCategory("Network")]
    [DynamicData(nameof(GetModelGroupFiles), DynamicDataSourceType.Method)]
    public async Task ModelUrlsShouldBeAccessible(string fileName)
```

Comment (weiyuanyue) [reply AIDevGallery.Tests/IntegrationTests/ModelUrlConnectivityTests.cs:71]: @copilot please make the change based on this comment

File: AIDevGallery.Utils/ModelUrl.cs (hunk 3 above)
Comment (weiyuanyue) [reply AIDevGallery.Utils/ModelUrl.cs:96]: nope. the search models on HF feature in AIDG builds URLs differently from the Build*Url methods and directly depends on ApiUrl, which therefore needs to remain public.
This PR only focuses on extracting and optimizing the URL builder for Model URL Connectivity Tests. The search feature is intentionally out of scope, as including it would introduce unnecessary changes

File: AIDevGallery.Utils/ModelUrl.cs (hunk 4 above)
Comment (weiyuanyue) [reply AIDevGallery.Utils/ModelUrl.cs]: Error CS0117 'ArgumentNullException' does not contain a definition for 'ThrowIfNullOrWhiteSpace'
AIDevGallery.Utils (netstandard2.0) :-D
